### ISSUE TITLE
Make regorule a service that is set up by the OPA operator

### ISSUE THREAD

**Issue by razvan** [stackabletech/regorule-operator#190]
## The problem

The regorule operator is in fact a service. It doesn't operate/manage anything. The OPA service is (by design) tightly coupled to the  regorule service.

In order for OPA to reach regorule, they both need to live in the same kubernetes namespace. **Currently with the kuttl tests and the kind cluster set-up apriori  this is not true because the regorule operator is installed before the OPA service is created.**

The workaround is for the kuttl tests to first remove the regorule operator from the default namespace and then reinstall it in the autogenerated namespace.

Another problem is that the trino-operator has a direct cargo dependency here. When this is implemented, the trino-operator would use a discovery config map to connect to regorule.

Additional problems with the current regorule service implementation:

* Cannot add data files to the bundles
* All policies in one bundle file
  * potential performance issue: invalidates all opa caches even for products that are not changed.
* Race condition between writing a bundle and serving it
* Several todos in code that we can get for free with nginx (tls, etag, etc.)

## Proposed solution

TL;DR: The opa-operator sets up an nginx service for bundle delivery and watches for `RegoBundle` resources. It builds bundles and uploads them to the nginx service where they are downloaded by OPA it's self.

More details:

* The opa-operator can set up a regorule bundle service based on nginx and configure the OPA servers with it. In the future, it bundles should be downloadable also from cloud-provides supported by OPA.
* Replace the regorule operator with nginx ~~+ upload module (https://www.nginx.com/resources/wiki/modules/upload/)~~. **Update**: No extra module is needed.
* OPA already supports nginx with tls, etag and authentication tokens. (no need to implement them ourselves)
* The OPA operator watches for `RegoBundle` custom resources, creates the bundles and uploads them to nginx.

**Comment by teozkr** [comment]
> In order for OPA to reach regorule, they both need to live in the same kubernetes namespace.

This isn't strictly true. `http://foo/` assumes will try to access the service `foo` in the pod's current namespace, but `http://foo.bar/` or `http://foo.bar.svc.cluster.local/` will both resolve `foo` in the namespace `bar`, regardless of the current namespace.

That said, it's still worth considering how we want to handle multitenancy in the long run...

> Another problem is that the trino-operator has a direct cargo dependency here. When this is implemented, the trino-operator would use a discovery config map to connect to regorule.

`ConfigMap` vs `RegoRule` CRD is completely unrelated to how the bundles are served.

> OPA already supports nginx with tls, etag and authentication tokens. (no need to implement them ourselves)

Do we need anything more than https://docs.rs/warp/latest/warp/struct.TlsServer.html for the TLS? Etag seems like overkill for v1 (and easy to add later, should the need actually arise), and authn seems fairly pointless as long as we're only serving static content.
